### PR TITLE
Support for np.quantile and np.nanquantile

### PR DESCRIPTION
**Pull request by rjenc29**
Initial commit for CI.

### CODE REVIEW

**Comment by rjenc29** [comment]
Un-WIP-ing as the tests have made it through.

The intention here is to add support for `np.quantile` and `np.nanquantile` as small extensions of `np.percentile` and `np.nanpercentile`

In passing, I added support for a few more edge cases like 0d inputs, and leveraged a few recent inventions like `np.asarray` to reduce code complexity and branching.  

In the process of testing, I found (and reported) a bug in numpy (added a reference and link in one of the tests, which uses float-coercion to bypass the bug).

Of note:

* The tests are arguably repetitive (as the 'same' underlying implementation is reached) and expensive.
* I advertised that complex dtypes are not supported - it would be possible to support them I think, but this would be more involved.
 
Anyway, I'd be grateful if you could let me know what you think in due course.  TA!

**Comment by aldanor** [comment]
Is there any chance to support `axis` argument?

**Comment by stuartarchibald** [comment]
@aldanor not in this PR I'm afraid. If `axis` support is important to you please open a new issue, thanks.

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -1080,11 +1080,33 @@ def _can_collect_percentiles(a, nan_mask, skip_nan):
...
-    if np.any(np.isnan(q)) or np.any(q < 0) or np.any(q > 100):
-        raise ValueError('Percentiles must be in the range [0,100]')
+def check_valid(q, q_upper_bound):
+    valid = True
+
+    # avoid expensive reductions where possible
+    if q.ndim == 1 and q.size < 10:
+        for i in range(q.size):
+            if q[i] < 0.0 or q[i] > q_upper_bound or np.isnan(q[i]):
+                valid = False
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/targets/arraymath.py]: Fixed in 7e2496f

**Comment by stuartarchibald** [comment]
@rjenc29 With view of getting this merged for 0.44 I pushed the 1 line fix to your branch. Hope that's ok? Thanks again for all your work on this. You'll be pleased to know there's a PR in the pipeline that's going to help with branch prune on freevar/global, which should make some of the contrived patterns that are necessary at present go away in future! 

Once CI passes this can be marked for merge.

**Pull request merged**